### PR TITLE
OAuth Authorize flow: remove intermediate screen

### PR DESCRIPTION
**Pull request by jsnajdr**
This PR removes the intermediate `/authorize` screen when OAuth-enabled Calypso doesn't have a `wpcom_token`, i.e., is logged out:

Screenshot 2021-05-04 at 9 52 47

Instead, we now navigate directly to `public-api.wordpress.com/oauth2/authorize`.

The login flow is now as follows:
- go to `http://calypso.localhost:3000/media?flags=oauth`
- Calypso discovers it's logged out and navigates to
```
https://public-api.wordpress.com/oauth2/authorize?redirect_to=http://calypso.localhost:3000/api/oauth/token?next=/media
```
(note how the `redirect_to` param navigates the flow to an "after login" `/api/oauth/token` page and how it gives it instructions where to navigate `next` after processing the login -- to the original `/media`)
- after approving the Calypso app (this additional step is further removed in D60952-code) you are redirected back to `/media`, this time successfully authorized.

Cc @cbauerman: Jetpack Cloud could benefit from a similar simplification: remove the `/connect` page, and do an implicit auth on backend.

### CODE REVIEW

File: client/auth/style.scss [hunk 1]
```diff
@@ -251,15 +251,6 @@
...
-.auth__welcome {
-	color: var( --color-text-inverted );
-}
-
-.auth__connect-intro {
-	width: 420px;
-	margin: 0 auto;
-}
-
 .auth__connect .button,
```
Comment (tyxla) [inline client/auth/style.scss]: We can also remove these, they're no longer used.

File: client/auth/controller.js [hunk 2]
```diff
@@ -18,54 +17,26 @@ import store from 'store';
...
+		return;
+	}
+
+	context.primary = <OAuthLogin />;
+	next();
+}
+
+// Store token into local storage
+export function storeToken( context ) {
+	if ( context.hash && context.hash.access_token ) {
```
Comment (tyxla) [inline client/auth/controller.js]: Nit: Now that we're touching this one, perhaps we could use optional chaining?

File: client/auth/controller.js [hunk 3]
```diff
@@ -18,54 +17,26 @@ import store from 'store';
...
+		store.set( 'wpcom_token', context.hash.access_token );
+	}
+
+	if ( context.hash && context.hash.expires_in ) {
```
Comment (tyxla) [inline client/auth/controller.js]: Nit: Now that we're touching this one, perhaps we could use optional chaining?

File: client/boot/common.js [hunk 4]
```diff
@@ -116,6 +116,27 @@ const setupContextMiddleware = ( reduxStore ) => {
 	} );
 };
 
+function authorizePath() {
+	// TODO: flags=oauth should be present only in dev-like environments
+	// where we need to preserve the flag on full reloads and redirects
+	const redirectUri = new URL( '/api/oauth/token', window.location );
+	redirectUri.search = new URLSearchParams( {
+		flags: 'oauth',
```
Comment (tyxla) [inline client/boot/common.js:124]: Feels so odd to hardcode that here. Should we wrap it in a check that does it only for the dev envs? DId you mean to do it as part of this PR?

File: client/boot/common.js [hunk 5]
```diff
@@ -116,6 +116,27 @@ const setupContextMiddleware = ( reduxStore ) => {
...
+		scope: 'global',
+		blog_id: 0,
+	} );
+
+	return authUri.toString();
+}
+
 const oauthTokenMiddleware = () => {
 	if ( config.isEnabled( 'oauth' ) ) {
 		const loggedOutRoutes = [
```
Comment (tyxla) [inline client/boot/common.js]: Should we remove `/authorize` from this list, too?

File: client/boot/common.js (hunk 4 above)
Comment (jsnajdr) [reply client/boot/common.js:124]: The `flags=oauth` query param was already hardcoded before this patch in the `getToken` function that stored the token and navigated to `/?flags=oauth`. This PR merely moves the code to a different location and adds the `next` query param that allows to configure the return route.

I plan to iterate further on this and eventually remove the query param. The goal is to preserve the `oauth` config flag on full reloads. And Jetpack Cloud already does a similar thing better: they have an `/oauth-override` endpoint that stores an OAuth-disabling flag to session storage and survives reloads of the tab: https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/jetpack/oauth-override.ts

**Comment by jsnajdr** [comment]
@tyxla your feedback was addressed: optional chaining, unused CSS styles, list of `loggedOutRoutes` -- all done.

**Comment by tyxla** [comment]
> @tyxla your feedback was addressed: optional chaining, unused CSS styles, list of `loggedOutRoutes` -- all done.

I don't see anything - perhaps you missed to push it to the branch?

**Comment by jsnajdr** [comment]
> perhaps you missed to push it to the branch?

oh yes, pushed now :+1:

**Review by tyxla — APPROVED** [review]
This looks good to me, code-wise. I haven't tested FWIW - so feel free to ship once you get a confirmation that it's working as expected.

**Pull request merged**